File: problems/0743-network-delay-time/typescript/solution.ts
class Heap<T> {
  private data: T[] = [];
  constructor(private less: (a: T, b: T) => boolean) {}

  size(): number {
    return this.data.length;
  }

  push(v: T): void {
    this.data.push(v);
    let i = this.data.length - 1;
    while (i > 0) {
      const p = (i - 1) >> 1;
      if (!this.less(this.data[i], this.data[p])) break;
      [this.data[p], this.data[i]] = [this.data[i], this.data[p]];
      i = p;
    }
  }

  pop(): T {
    const top = this.data[0];
    const last = this.data.pop();
    if (last !== undefined && this.data.length > 0) {
      this.data[0] = last;
      let i = 0;
      const n = this.data.length;
      while (true) {
        let best = i;
        const l = 2 * i + 1;
        const r = 2 * i + 2;
        if (l < n && this.less(this.data[l], this.data[best])) best = l;
        if (r < n && this.less(this.data[r], this.data[best])) best = r;
        if (best === i) break;
        [this.data[best], this.data[i]] = [this.data[i], this.data[best]];
        i = best;
      }
    }
    return top;
  }
}

// Network Delay Time — https://leetcode.com/problems/network-delay-time/
export function networkDelayTime(times: number[][], n: number, k: number): number {
  const graph = new Map<number, [number, number][]>();
  for (const [u, v, w] of times) {
    const list = graph.get(u) ?? [];
    list.push([v, w]);
    graph.set(u, list);
  }

  const dist = new Map<number, number>();
  const heap = new Heap<[number, number]>((a, b) => a[0] < b[0]); // (dist, node)
  heap.push([0, k]);
  while (heap.size() > 0) {
    const [d, node] = heap.pop();
    if (dist.has(node)) continue;
    dist.set(node, d);
    for (const [nei, w] of graph.get(node) ?? []) {
      if (!dist.has(nei)) heap.push([d + w, nei]);
    }
  }

  if (dist.size < n) return -1;
  return Math.max(...dist.values());
}

File: problems/0743-network-delay-time/go/solution.go
package solution

import "container/heap"

type state struct{ dist, node int }

type stateHeap []state

func (h stateHeap) Len() int           { return len(h) }
func (h stateHeap) Less(i, j int) bool { return h[i].dist < h[j].dist }
func (h stateHeap) Swap(i, j int)      { h[i], h[j] = h[j], h[i] }
func (h *stateHeap) Push(x any)        { *h = append(*h, x.(state)) }
func (h *stateHeap) Pop() any {
	o := *h
	n := len(o)
	x := o[n-1]
	*h = o[:n-1]
	return x
}

// Network Delay Time — https://leetcode.com/problems/network-delay-time/
func networkDelayTime(times [][]int, n, k int) int {
	graph := map[int][][2]int{}
	for _, t := range times {
		graph[t[0]] = append(graph[t[0]], [2]int{t[1], t[2]})
	}

	dist := map[int]int{}
	h := &stateHeap{{0, k}}
	for h.Len() > 0 {
		s := heap.Pop(h).(state)
		if _, ok := dist[s.node]; ok {
			continue
		}
		dist[s.node] = s.dist
		for _, e := range graph[s.node] {
			if _, ok := dist[e[0]]; !ok {
				heap.Push(h, state{s.dist + e[1], e[0]})
			}
		}
	}

	if len(dist) < n {
		return -1
	}
	maxDist := 0
	for _, d := range dist {
		if d > maxDist {
			maxDist = d
		}
	}
	return maxDist
}
